### PR TITLE
Add media segment skipping

### PR DESCRIPTION
**Pull request by thornbill**
**Changes**
* Adds support for skipping media segments

Needs cleanup + testing so marking as a draft for now

**Issues**
N/A

### CODE REVIEW

**Comment by viown** [comment]
Just some things I noted when testing:

- If I have "Skip" enabled and seek to a duration between the segment, it'll immediately take me to the end of the segment. If there is something there I wanted to take a look at I wouldn't be able to get to it. Maybe it should only skip when it hits the start time of the segment or perhaps avoid skipping altogether when the user manually seeks?
- When it's about to skip a segment, it can look like the stream is frozen (this is amplified when transcoding). Having some kind of visual indicator (like the one for syncplay) would probably work well here.

**Comment by solidsnake1298** [comment]
From my initial testing, this reliably skips intros, but it has not skipped outros or previews.

I'm going to create some test media so I can test the other segment types.

**Comment by viown** [comment]
> From my initial testing, this reliably skips intros, but it has not skipped outros or previews.
> 
> I'm going to create some test media so I can test the other segment types.

It looks like all segments after the first will not be skipped because `segment` is not updated here:

https://github.com/jellyfin/jellyfin-web/blob/91de060231bd9b1ef5ce094568bc03a47dac17ad/src/apps/stable/features/playback/utils/mediaSegmentManager.ts#L47

**Comment by solidsnake1298** [comment]
The latest commit is skipping intro, outro, and preview segments reliably.  I haven't gotten a chance to create test media with other segment types, but I don't see why those wouldn't also skip.

**Comment by solidsnake1298** [comment]
I created some test media for recap and commercial segments, which were also skipped.

File: src/apps/stable/features/playback/utils/mediaSegmentManager.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                this.performAction(currentSegmentDetails.segment);
+                this.lastIndex = currentSegmentDetails.index;
+            }
+        }
+    }
+
+    private debouncedOnTimeUpdate = throttle(this.handleSegmentActions, 500, { leading: true });
+
+    private async fetchMediaSegments(api: Api, itemId: string, includeSegmentTypes: MediaSegmentType[]) {
+        // FIXME: Replace with SDK getMediaSegmentsApi function when available in stable
```
[the remark this answers is not in the document]
Comment (thornbill) [reply src/apps/stable/features/playback/utils/mediaSegmentManager.ts:25]: This is dependent on the media segment api being available in the stable sdk release.

**Comment by thornbill** [comment]
These are great points.

>     * If I have "Skip" enabled and seek to a duration between the segment, it'll immediately take me to the end of the segment. If there is something there I wanted to take a look at I wouldn't be able to get to it. Maybe it should only skip when it hits the start time of the segment or perhaps avoid skipping altogether when the user manually seeks?

Good idea. I think this makes sense as an enhancement. We should align on the desired behavior with other client implementations (Android TV).
 
>     * When it's about to skip a segment, it can look like the stream is frozen (this is amplified when transcoding). Having some kind of visual indicator (like the one for syncplay) would probably work well here.

I can definitely see this on my unstable server where I have assigned very limited resources. I wonder if just a toast message when skipping starts would be good enough for now. :thinking:

**Comment by nielsvanvelzen** [comment]
> > * If I have "Skip" enabled and seek to a duration between the segment, it'll immediately take me to the end of the segment. If there is something there I wanted to take a look at I wouldn't be able to get to it. Maybe it should only skip when it hits the start time of the segment or perhaps avoid skipping altogether when the user manually seeks?
> 
> Good idea. I think this makes sense as an enhancement. We should align on the desired behavior with other client implementations (Android TV).

This is how I defined the behavior for the skip action in ATV, from https://github.com/jellyfin/jellyfin-androidtv/blob/master/app/src/main/java/org/jellyfin/androidtv/ui/playback/segment/MediaSegmentAction.kt#L15-L16

```
	/**
	 * Seek to the end of this segment (endTicks). If the duration of this segment is shorter than 1 second it should do nothing to avoid
	 * lagg. The skip action will only execute when playing over the segment start, not when seeking into the segment block.
	 */
	SKIP,

	/**
	 * Ask the user if they want to skip this segment. When the user agrees this behaves like [SKIP]. Confirmation should only be asked for
	 * segments with a duration of at least 3 seconds to avoid UI flickering.
	 */
	ASK_TO_SKIP,

	/**
	 * Mute the audio volume until the end of this segment. The volume level should be restored to the state when this segment started. If
	 * the volume is manually changed by the user during this segment it should no longer restore volume itself. The mute action will always
	 * execute when the segment block is active.
	 */
	MUTE,
```

**Comment by rlauuzo** [comment]
I suggest adding a configurable buffer time (e.g., 2 seconds) at the start and end of segments to prevent accidental content skipping.

**Review by dmitrylyzo — COMMENTED** [review]
Haven't tested - just reviewed. 😅

Am I right, it will skip the "outro" even we seek to it?

**Comment by thornbill** [comment]
> Am I right, it will skip the "outro" even we seek to it?

I just pushed a change that should bring the behavior inline with what Niels implemented for Android TV... I hope. 🤞 

If you seek forward into a segment it will take the action as normal. If your previous player time is later than the start time of the segment no action is taken.

I made this behavior specific to the skip action because I think for others (prompting) it might still make sense to invoke the action. 🤷‍♂️

**Comment by nielsvanvelzen** [comment]
Skipping seems to work mostly as I'd expect except for one issue that I encountered. When I choose to skip outros and manually seek into an outro segment, it will be skipped. My expectation would be that manually seeking into it will not trigger the skip.

Otherwise it doesn't have some bugs that the current ATV implementation has, so that's great. (In ATV skipping to the end of the video will close the player instead of going to the next queue item :p)

**Comment by thornbill** [comment]
Yeah using the current time base approach we can only detect if we skip back into a segment not forward... I don't think we currently have any better way to detect when we have skipped unless we check if the difference between playtime updates is greater than some threshold. :man_shrugging:

**Comment by viown** [comment]
> Yeah using the current time base approach we can only detect if we skip back into a segment not forward... I don't think we currently have any better way to detect when we have skipped unless we check if the difference between playtime updates is greater than some threshold. 🤷‍♂️

Perhaps we could only trigger the skip if the current time is close to `StartTicks` by some difference (<1s)?

**Comment by dmitrylyzo** [comment]
We will probably perform bigger steps with increased playback rate, and difference will depend on that.

The most reliable way is to add some extra flag to the `timeupdate` event, which is on when we set the current time of the player.

**Comment by thornbill** [comment]
I was thinking we could add an event on seeking in the playback manager and use that to ignore the segment if the next time update falls in one...

My question is if the current implementation is "good enough" for the initial PR or if this should be added before merging. 🤷‍♂️ I'm hoping to get this in for the next unstable build.

**Comment by nielsvanvelzen** [comment]
IMO current state is good enough to merge and we can improve upon it in 10.10.z releases.

**Review by nielsvanvelzen — APPROVED** [review]

**Comment by pimw1** [comment]
I've installed the "Chapter Segments"  plugin and succesfully ran the "media segment scan" . For example, it was succesful for the below series episode.

`[2024-10-14 15:02:52.631 +02:00] [INF] [24] Jellyfin.Server.Implementations.MediaSegments.MediaSegmentManager: Media Segment provider "Chapter Segments Provider" found 1 for "/data/media/tv/Criminal Minds/Season 03/Criminal Minds (2005) - S03E12 - 3rd Life [DSNP WEBDL-1080p][EAC3 5.1][h264]-playWEB.mkv"`

When using https://cdb680bb.jellyfin-web.pages.dev/ to play this episode, i see the intro is recognized in the timeline

image

However, i don't get popup to skip the intro when the series progresses into that segment. How should this look like / behave?

**Comment by dmitrylyzo** [comment]
> How should this look like / behave?

You need to select an action for the segment type in `User settings / Playback`. The action is applied without a prompt. In the future, there will be "Ask to skip".

**Comment by pimw1** [comment]
Ah got it, thanks. It works as expected. I also notice that the transcoding is apparently taking it into account, since i don't notice the usual delay (which i would notice when manually skipping through the media). Nice!

**Comment by pimw1** [comment]
Just a thought: introduce a global setting for skipping behaviour per segment type, which users can overwrite when they choose to. Reason: i don't see my non-techy family members enabling the skip behaviour.

**Comment by pimw1** [comment]
Upon further testing, intro skipping sometimes does cause a tempory freeze due to transcoding. For a future update, it would be nice if transcoding would skip the intro anyway if the user setting is set to skip.

**Pull request merged**